### DISCUSSION TITLE
Unit Test

### DISCUSSION THREAD

**Discussion by tiagosouza488** [Kamva/mgm#71] [Q&A] [answered]
I looked through all the examples in the repository but I didn't find a way to use mocks. Is something like go.mongodb.org/mongo-driver/mongo/integration/mtest possible?

`
func TestFindOne(t *testing.T) {
	mt := mtest.New(t, mtest.NewOptions().ClientType(mtest.Mock))
	defer mt.Close()

	mt.Run("success", func(mt *mtest.T) {
		userCollection = mt.Coll
		expectedUser := user{
			ID:    primitive.NewObjectID(),
			Name:  "john",
			Email: "john.doe@test.com",
		}

		mt.AddMockResponses(mtest.CreateCursorResponse(1, "foo.bar", mtest.FirstBatch, bson.D{
			{"_id", expectedUser.ID},
			{"name", expectedUser.Name},
			{"email", expectedUser.Email},
		}))
		userResponse, err := getFromID(expectedUser.ID)
		assert.Nil(t, err)
		assert.Equal(t, &expectedUser, userResponse)
	})
}
`

**Comment by mehran-prs** [comment, marked as answer]
Yes, I think you can use `go.mongodb.org/mongo-driver/mongo/integration/mtest` for mgm as well, because it uses `mongo-driver` library under the hood.
For example, when you call the `FindByID` method on an mgm collection, it uses `FindOne` method of the mongo-driver's `Collection` under the hood, so you can mock `FinOne` and then test it. 
mgm's `Collection` embeds mongo-driver's `Collection` actually, so you can mock all of the mongo-driver methods as well.

I'm just unsure whether we need to provided the mocked mongo-driver's `Collection` to mgm's `Collection` or it let us create the mgm Collection from mocked DB instance. If it needs, we have two solutions:

- Add mgm `Collection` constructor that gets the mocked mongo-driver's `Collection` as a param.
- Update mgm `Collection` manually, like this:
```go
c:=mgm.CollectionByName("test")
c.Collection=mt.Coll // assign mocked collection as a mongo-driver collection to mgm Collection
```
